/*
* @license Apache-2.0
*
* Copyright (c) 2023 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import srotmg = require( './index' );


// TESTS //

// The function returns a Float32Array...
{
	srotmg( 0.0, 1.0, 2.0, 3.0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the function is provided an argument which is not a number...
{
	srotmg( true, 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg( false, 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg( null, 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg( undefined, 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg( '5', 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg( [], 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg( {}, 1.0, 2.0, 3.0 ); // $ExpectError

	srotmg( 0.0, true, 2.0, 3.0 ); // $ExpectError
	srotmg( 0.0, false, 2.0, 3.0 ); // $ExpectError
	srotmg( 0.0, null, 2.0, 3.0 ); // $ExpectError
	srotmg( 0.0, undefined, 2.0, 3.0 ); // $ExpectError
	srotmg( 0.0, '5', 2.0, 3.0 ); // $ExpectError
	srotmg( 0.0, [], 2.0, 3.0 ); // $ExpectError
	srotmg( 0.0, {}, 2.0, 3.0 ); // $ExpectError

    srotmg( 0.0, 1.0, true, 3.0 ); // $ExpectError
	srotmg( 0.0, 1.0, false, 3.0 ); // $ExpectError
	srotmg( 0.0, 1.0, null, 3.0 ); // $ExpectError
	srotmg( 0.0, 1.0, undefined, 3.0 ); // $ExpectError
	srotmg( 0.0, 1.0, '5', 3.0 ); // $ExpectError
	srotmg( 0.0, 1.0, [], 3.0 ); // $ExpectError
	srotmg( 0.0, 1.0, {}, 3.0 ); // $ExpectError

    srotmg( 0.0, 1.0, 2.0, true ); // $ExpectError
	srotmg( 0.0, 1.0, 2.0, false ); // $ExpectError
	srotmg( 0.0, 1.0, 2.0, null ); // $ExpectError
	srotmg( 0.0, 1.0, 2.0, undefined ); // $ExpectError
	srotmg( 0.0, 1.0, 2.0, '5' ); // $ExpectError
	srotmg( 0.0, 1.0, 2.0, [] ); // $ExpectError
	srotmg( 0.0, 1.0, 2.0, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	srotmg(); // $ExpectError
	srotmg( 0.0 ); // $ExpectError
	srotmg( 0.0, 1.0 ); // $ExpectError
    srotmg( 0.0, 1.0, 2.0 ); // $ExpectError
    srotmg( 0.0, 1.0, 2.0, 3.0, 4.0 ); // $ExpectError
}

// Attached to the main export is an `assign` method which returns a Float32Array...
{
	const out = new Float32Array( 5 );

	srotmg.assign( 0.0, 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectType Float32Array
}

// The compiler throws an error if the `assign` method is provided a first or second argument which is not a number...
{
	const out = new Float32Array( 5 );

	srotmg.assign( true, 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( false, 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( null, 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( undefined, 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( '5', 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( [], 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( {}, 2.0, 3.0, 4.0, out, 1, 0 ); // $ExpectError

	srotmg.assign( 0.0, true, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, false, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, null, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, undefined, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, '5', 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, [], 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, {}, 3.0, 4.0, out, 1, 0 ); // $ExpectError

    srotmg.assign( 0.0, 2.0 ,true, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, 2.0 ,false, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, 2.0 ,null, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, 2.0 ,undefined, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, 2.0 ,'5', 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, 2.0 ,[], 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, 2.0 ,{}, 4.0, out, 1, 0 ); // $ExpectError

    srotmg.assign( 0.0, true, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, false, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, null, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, undefined, 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, '5', 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, [], 3.0, 4.0, out, 1, 0 ); // $ExpectError
	srotmg.assign( 0.0, {}, 3.0, 4.0, out, 1, 0 ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided a fourth argument which is not a number...
{
	const out = new Float32Array( 5 );

	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, '5', 0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, true, 0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, false, 0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, null, 0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, [], 0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, {}, 0 ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided a fifth argument which is not a number...
{
	const out = new Float32Array( 5 );

	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, 1, '5' ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, 1, true ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, 1, false ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, 1, null ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, 1, [] ); // $ExpectError
	srotmg.assign( 1.0, 2.0, 3.0, 4.0, out, 1, {} ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided an unsupported number of arguments...
{
	const out = new Float32Array( 5 );

	srotmg.assign(); // $ExpectError
	srotmg.assign( 1.0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0 ); // $ExpectError
    srotmg.assign( 1.0, 2.0, 3.0 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, out ); // $ExpectError
    srotmg.assign( 1.0, 2.0, 3.0, out ); // $ExpectError
	srotmg.assign( 1.0, 2.0, out, 1 ); // $ExpectError
    srotmg.assign( 1.0, 2.0, 3.0, out, 1 ); // $ExpectError
	srotmg.assign( 1.0, 2.0, out, 1, 0, 1 ); // $ExpectError
    srotmg.assign( 1.0, 2.0, 3.0, out, 1, 0, 1 ); // $ExpectError
}